src/Microsoft.VisualStudio.Threading/JoinableTask`1.cs

```diff
@@ -47,10 +47,27 @@ internal JoinableTask(JoinableTaskFactory owner, bool synchronouslyBlocking, Joi
         /// </summary>
         /// <param name="cancellationToken">A cancellation token that will exit this method before the task is completed.</param>
         /// <returns>A task that completes after the asynchronous operation completes and the join is reverted, with the result of the operation.</returns>
-        public new async Task<T> JoinAsync(CancellationToken cancellationToken = default(CancellationToken))
+        public new Task<T> JoinAsync(CancellationToken cancellationToken = default(CancellationToken))
         {
-            await base.JoinAsync(cancellationToken).ConfigureAwait(AwaitShouldCaptureSyncContext);
-            return await this.Task.ConfigureAwait(AwaitShouldCaptureSyncContext);
+            Task baseJoin = base.JoinAsync(cancellationToken);
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return System.Threading.Tasks.Task.FromCanceled<T>(cancellationToken);
+            }
```

Comment:
The `base.JoinAsync` method already returns a canceled Task if `cancellationToken` is canceled. This double-check actually results in the caller receiving the exception *before* `baseJoin.IsCompleted`, so the dis-join may not have completed yet, which seems bad to me.

Reply:
➡️ I cleaned this one up